2FA with WebAuthn.

alesstimec opened the pull request.

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/candid/146/

rogpeppe commented on store/testing/meeting.go:
```diff
@@ -34,7 +34,9 @@ func (s *MeetingSuite) SetUpTest(c *gc.C) {
 
 func (s *MeetingSuite) TearDownTest(c *gc.C) {
-	s.close()
+	if s.close != nil {
```
When might this be nil?

mhilton replied:
When it wasn't set up because the database connection failed.

rogpeppe commented on store/testing/store.go:
```diff
@@ -41,7 +41,9 @@ func (s *StoreSuite) SetUpTest(c *gc.C) {
 
 func (s *StoreSuite) TearDownTest(c *gc.C) {
-	s.close()
+	if s.close != nil {
```
Similar here.

rogpeppe commented on store/mgostore/bakery_test.go:
```diff
@@ -4,22 +4,39 @@
+func (s *bakerySuite) SetUpTest(c *gc.C) {
+	var err error
+	s.db, err = mgotest.New()
+	c.Logf("ERR: %#v", err)
```
debugging remnant?

mhilton commented:
:shipit:

jujugui commented:
Status: merge request accepted. Url: http://ci-gce.theblues.io:8080/job/candid-merge

jujugui commented:
Build failed: Attempt to land pull request failed
        build url: http://ci-gce.theblues.io:8080/job/candid-merge/64

alesstimec committed:
2FA with WebAuthn.

dependabot[bot] committed:
build(deps): bump github.com/google/uuid from 1.2.0 to 1.3.0

Bumps [github.com/google/uuid](https://github.com/google/uuid) from 1.2.0 to 1.3.0.
- [Release notes](https://github.com/google/uuid/releases)
- [Commits](https://github.com/google/uuid/compare/v1.2.0...v1.3.0)

---
updated-dependencies:
- dependency-name: github.com/google/uuid
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

mhilton committed:
Merge pull request #69 from canonical/dependabot/go_modules/github.com/google/uuid-1.3.0

build(deps): bump github.com/google/uuid from 1.2.0 to 1.3.0

mhilton commented on idp/idputil/idputil.go:
```diff
@@ -214,6 +218,17 @@ func HandleLoginForm(
 	return nil, errgo.Mask(tmpl.ExecuteTemplate(w, "login-form", data))
 }
 
+func PresentForm(ctx context.Context, w http.ResponseWriter, templateName string, params interface{}, t *template.Template) error {
```
Please add a doc comment.

alesstimec replied:
added

mhilton replied:
where?

alesstimec replied:
removed the PresentForm function

mhilton commented on idp/idputil/idputil.go:
```diff
@@ -254,3 +269,308 @@ func CookiePathRelativeToLocation(cookiePath, location string, skipLocation bool
+
+// JsonResponse mashals the provided to JSON and writes it
+// to the provided response writer.
+func JsonResponse(w http.ResponseWriter, data interface{}, c int) {
```
httprequest.WriteJSON (or similar does this).

alesstimec replied:
done

mhilton commented on idp/idputil/idputil.go:
```diff
@@ -254,3 +269,308 @@ func CookiePathRelativeToLocation(cookiePath, location string, skipLocation bool
+	}
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(c)
+	fmt.Fprintf(w, "%s", dj)
```
Why are you formatting a slice of bytes as a string to write to the writer? it seems like a very inefficient way to do `w.Write(dj)` note the contract for Writer requires the write to either succeed fully, or fail with a terminal error.

mhilton commented on idp/idputil/idputil.go:
```diff
@@ -254,3 +269,308 @@ func CookiePathRelativeToLocation(cookiePath, location string, skipLocation bool
+
+// WebAuthnID implements the webauthn.User interface.
+func (u *webauthnUser) WebAuthnID() []byte {
+	return []byte(u.Identity.Username)
```
depending on what this is used for the ProviderID might be a better choice.

alesstimec replied:
done

mhilton commented on idp/idputil/idputil.go:
```diff
@@ -254,3 +269,308 @@ func CookiePathRelativeToLocation(cookiePath, location string, skipLocation bool
+	fmt.Fprintf(w, "%s", dj)
+}
+
+// WebAuthnUserFromIdentity converts the provided store Identity to
```
I think I would be inclined to put the webauthn stuff into it's own file, or possibly even package.

alesstimec replied:
done

mhilton commented on templates/mfa-login:
```diff
@@ -0,0 +1,128 @@
+    }
+  </script>
+</body>
+</html>
```
please add a trailing newline

alesstimec replied:
done

mhilton commented on templates/mfa-register:
```diff
@@ -0,0 +1,133 @@
+</script>
+
+</body>
+</html>
```
and here.

alesstimec replied:
done

mhilton reviewed (changes requested):
please put the webauthn related code into it's own package.

mhilton committed:
internal/discharge: include URL in invalid return_to errors

mhilton committed:
Merge pull request #72 from mhilton/209-invalid-return-to-url

internal/discharge: include URL in invalid return_to errors

mhilton commented on store/store.go:
```diff
@@ -235,3 +248,23 @@ type Identity struct {
+type MFACredential struct {
+	// Username contains the username of the user to which the
+	// multi-factor credential belongs.
+	Username string
```
It would be better to use a field that is fixed for this, under some circumstances the username can change. ProviderID is probably  good choice.

alesstimec replied:
i proposed using identity.ID...

mhilton replied:
The only thing about using identity.ID is that you will always have to look that up in the database, whereas ProviderID can normally be derived from information provided by the identity provider. In this case it probably doesn't matter.

alesstimec replied:
changed to ProviderID

jcsackett commented on idp/idputil/mfa/webauthn.go:
```diff
@@ -0,0 +1,407 @@
+const MFACookiePath = "/login"
+
+// LoginState holds the state of the current multi-factor
+// authentication loging process.
```
s/loging/logging

mhilton replied:
probably actually "login"

jcsackett commented on idp/static/static.go:
```diff
@@ -196,11 +237,122 @@ func (idp *identityProvider) Handle(ctx context.Context, w http.ResponseWriter,
+				}
+				note := ""
+				if len(userCredentials) == 0 {
+					note = "Identity provide requires MFA. Please register a security key."
```
s/provide/provider/

mhilton commented on cmd/candidsrv/config.yaml:
```diff
@@ -15,6 +15,10 @@ identity-providers:
 - type: static
   name: static
   description: Default identity provider
+  required-2fa: true
```
We will probably want some charm changes as a follow-up.

mhilton commented on go.mod:
```diff
@@ -1,12 +1,30 @@
 module github.com/canonical/candid
 
 require (
+	github.com/bombsimon/wsl v1.2.8 // indirect
```
This is quite the increase in dependencies, have you performed a security audit of the new packages?

alesstimec replied:
not yet, but i should do..

mhilton commented on idp/ldap/ldap.go:
```diff
@@ -79,6 +79,9 @@ type Params struct {
 	Password string `yaml:"password"`
 
+	// Required2FA indicates if this provider requires the user of 2FA
+	Required2FA bool `yaml:"required-2fa"`
```
Should we be consistent about the terms MFA & 2FA, or do they really mean different things?

mhilton replied:
presumably there's some implementation of this that will come in a follow-up?

alesstimec replied:
yes, of course

mhilton commented on idp/static/static.go:
```diff
@@ -63,6 +68,15 @@ type Params struct {
+	// Required2FA indicates if this provider requires the user of 2FA
+	Required2FA bool `yaml:"required-2fa"`
+
+	RPDisplayName string `yaml:"rp-display-name"`
```
Please add doc-comments to all of these fields

alesstimec replied:
done

mhilton commented on idp/static/static.go:
```diff
@@ -96,16 +110,34 @@ func NewIdentityProvider(p Params) idp.IdentityProvider {
 		params:         p,
 		matchEmailAddr: matchEmailAddr,
 	}
+	idp.authenticator = &mfa.MultiFactorAuthenticator{
```
Presumably this should be protected by an `if auth != nil`. Otherwise you'll add unusable authenticator to the idp.

alesstimec replied:
done

mhilton commented on idp/static/static.go:
```diff
@@ -174,15 +212,18 @@ func (idp *identityProvider) GetGroups(ctx context.Context, identity *store.Iden
 	return []string{}, nil
 }
 
+type message struct {
```
Even though this isn't exported it isn't immediately obvious what this is for, please add a doc-comment.

alesstimec replied:
added

mhilton commented on idp/static/static.go:
```diff
@@ -63,6 +68,15 @@ type Params struct {
 	MatchEmailAddr string `yaml:"match-email-addr"`
+
+	// Required2FA indicates if this provider requires the user of 2FA
+	Required2FA bool `yaml:"required-2fa"`
```
Why is this in the past tense? shouldn't it be `require-2fa`?

alesstimec replied:
changed

mhilton commented on idp/static/static.go:
```diff
@@ -196,11 +237,122 @@ func (idp *identityProvider) Handle(ctx context.Context, w http.ResponseWriter,
 		}
 		if id != nil {
-			idp.initParams.VisitCompleter.RedirectSuccess(ctx, w, req, ls.ReturnTo, ls.State, id)
+			mfaLoginState := mfa.MFALoginState{
```
I'd suggest having something like:
```
if !idp.params.Required2FA {
    idp.initParams.VisitCompleter.RedirectSuccess(ctx, w, req, ls.ReturnTo, ls.State, id)
}
```

up here to save yourself an indent level (and some unnecessary cookie writing)

alesstimec replied:
done

mhilton commented on idp/static/static.go:
```diff
@@ -196,11 +237,122 @@ func (idp *identityProvider) Handle(ctx context.Context, w http.ResponseWriter,
+				idp.initParams.VisitCompleter.RedirectSuccess(ctx, w, req, ls.ReturnTo, ls.State, id)
+			}
+		}
+	case "/remove":
```
These parts look like they could be put into a shared handler?

mhilton commented on idp/static/static.go:
```diff
@@ -196,11 +237,122 @@ func (idp *identityProvider) Handle(ctx context.Context, w http.ResponseWriter,
 	}
 }
 
+type Credential struct {
```
Please add doc-comments.

alesstimec replied:
moved to the mfa package and added a comment

mhilton commented on idp/static/static.go:
```diff
@@ -219,6 +371,11 @@ func (idp *identityProvider) loginUser(ctx context.Context, user, password strin
 			if err != nil {
 				return nil, errgo.Mask(err)
 			}
+			err = idp.initParams.Store.Identity(ctx, id)
```
It would be better if we didn't have to look up users here if we can avoid it.

alesstimec replied:
changed.. since we're using the ProviderID now.. though we still have to look up the credentials..

mhilton commented on idp/static/static.go:
```diff
@@ -196,11 +237,122 @@ func (idp *identityProvider) Handle(ctx context.Context, w http.ResponseWriter,
+	RemoveURL string
+}
+
+type mfaParams struct {
```
This looks like it should belong in the `mfa` package.

alesstimec replied:
yes.. moved

mhilton commented on idp/static/static.go:
```diff
@@ -196,11 +237,122 @@ func (idp *identityProvider) Handle(ctx context.Context, w http.ResponseWriter,
+	Credentials        []Credential
+}
+
+type webauthnCredentialAssertion struct {
```
as does this.

alesstimec replied:
removed

mhilton commented on store/store.go:
```diff
@@ -146,6 +146,19 @@ type Store interface {
+	UserMFACredentials(ctx context.Context, id string) ([]MFACredential, error)
+
+	// IncrementMFACredentialSignCount increments the multi-factor credential sign count.
+	IncrementMFACredentialSignCount(ctx context.Context, cred MFACredential) error
```
Does this interface really need the whole credential? won't the user and cred name suffice?

alesstimec replied:
identity ID and credential name, yes.. but both are contained in the MFACredential..

alesstimec replied:
changed

mhilton commented on store/store.go:
```diff
@@ -146,6 +146,19 @@ type Store interface {
+
+	// RemoveMFACredential removes the multi-factor credential with the
+	// specified username and credential name.
+	RemoveMFACredential(ctx context.Context, cred MFACredential) error
```
Does this interface really need the whole credential? won't the user and cred name suffice?

alesstimec replied:
identity ID and credential name, yes.. but both are contained in the MFACredential..

alesstimec replied:
changed

mhilton commented on store/store.go:
```diff
@@ -146,6 +146,19 @@ type Store interface {
 	IdentityCounts(ctx context.Context) (map[string]int, error)
+
+	// AddMFACredential stores the specified multi-factor credential.
+	AddMFACredential(ctx context.Context, cred MFACredential) error
```
I'm in two minds about adding to the store interface in this way. The intention was that, in general, the identity providers would use the ProviderInfo field to store provider-specific information about an identity. Here however it's not so clear-cut and maybe this is sufficiently complex to warrant new store interfaces.

However, if this really does need a new store interface we should probably move the bulk of the MFA code to be part of the main application, rather than hidden in an IDP.

mhilton commented on idp/idputil/mfa/webauthn.go:
```diff
@@ -0,0 +1,407 @@
+
+// WebAuthnID implements the webauthn.User interface.
+func (u *webauthnUser) WebAuthnID() []byte {
+	return []byte(u.Identity.ID)
```
I suspect ProviderID would be better for this. The ID field is an internal-only idea. Theoretically it can change if we do a database migration, for instance.

alesstimec replied:
yes.. changed

mhilton reviewed (commented):
I think I'd like to take another look at this before it lands.

The amount of MFA code that is in the identity provider surprised me, I wonder whether a better approach would be for the IDP to create the initial MFA state and then forward to a shared facility that deals with MFA for the whole application. It seems silly to add the same 6 endpoints to every IDP handler that requires MFA.

alesstimec committed:
Standalone MFA handler.

mhilton commented on cmd/candidsrv/config.yaml:
```diff
@@ -11,10 +11,14 @@ max-mgo-sessions: 300
 request-timeout: 2s
 private-addr: localhost
 admin-agent-public-key: iKp24EL2Aj9MQfRkpwzp7rz7Zf3QZsEzWGpoWT3OK2w=
+rp-display-name: Candid
```
can we name these mfa-rp-display-name etc. The problem with RP is that the term occurs in pretty much any authentication protocol so we will probably want another RP in a different context in future.

alesstimec replied:
done

mhilton commented on cmd/candidsrv/main.go:
```diff
@@ -178,7 +186,6 @@ func serveIdentity(conf *config.Config, params candid.ServerParams) error {
 		Handler:   server,
 		TLSConfig: conf.TLSConfig(),
 	}
-	fmt.Println("START")
```
I've just remembered what this is for. We use the START message to detect that the service is working. Please put it back.

alesstimec replied:
done

mhilton commented on config/config.go:
```diff
@@ -17,6 +17,7 @@ import (
 	"gopkg.in/yaml.v2"
 
 	"github.com/canonical/candid/idp"
+	"github.com/canonical/candid/internal/mfa"
```
I think it might be better if you passed a config into the mfa package rather than create an mfa Authenticator in this package. unless that causes an unexpected problem.

alesstimec replied:
done

mhilton commented on idp/idp.go:
```diff
@@ -43,6 +43,9 @@ type VisitCompleter interface {
 	RedirectSuccess(ctx context.Context, w http.ResponseWriter, req *http.Request, returnTo, state string, id *store.Identity)
+
+	// RedirectMFA redirects the user to the multi-factor login flow.
+	RedirectMFA(ctx context.Context, w http.ResponseWriter, req *http.Request, state, mfaState string)
```
shouldn't this also include the authenticated identity in the signature?

alesstimec replied:
no.. the identity (well, the provider ID) is in the mfa state..

mhilton commented on store/memstore/credential.go:
```diff
@@ -0,0 +1,72 @@
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	key := fmt.Sprintf("%s-%s", providerID, name)
```
why not use your mfaCredentialKey function?

alesstimec replied:
done

mhilton reviewed (approved):
This is looking really nice, I have a few suggestions but this can land.

mhilton commented on idp/static/static.go:
```diff
@@ -194,9 +197,16 @@ func (idp *identityProvider) Handle(ctx context.Context, w http.ResponseWriter,
-		if id != nil {
-			idp.initParams.VisitCompleter.RedirectSuccess(ctx, w, req, ls.ReturnTo, ls.State, id)
+		if id != nil && idp.params.Require2FA {
+			mfaState, err := idp.initParams.SetMFAStateProviderID(w, string(id.ProviderID))
```
could this part be done in RedirectMFA? to save every IDP from having to do the same thing?

alesstimec replied:
done

mhilton reviewed (approved):
one more thought, feel free to ignore
